### PR TITLE
docs: ipns

### PR DESCRIPTION
**Pull request by vasco-santos**
Specs for IPNS implementation

### CODE REVIEW

**Comment by vasco-santos** [comment]
cc @Stebalien

**Comment by vasco-santos** [comment]
Spec updated with the interop changes

File: naming/README.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+An IPNS record is a data structure composed by:
+
+- 1. **Value** (bytes)
+- 2. **Validity** (bytes)
+- 3. **Validity Type** (uint64)
+  - Allows us to define if the record is valid under certain conditions
+  - Only supporting expiration date with `validityType = 0` for now
+- 4. **Signature** (bytes)
+  - Concatenate value, validity field and validity type
+  - Sign the concatenation result with the provided private key
```
[the remark this answers is not in the document]
Comment (vasco-santos) [reply naming/README.md]: Yes, totally agree, we need to change the signature logic! 

We should rethink this once we make changes to the record definition.

File: naming/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+**Note**: validity must be with nanoseconds precision.
+
+#### Local record
+
+This record is stored in the peer's repo datastore and contains the **latest** version of the IPNS record published by the provided key. This record is useful for republishing, as well as tracking the sequence number.
+
+**Key format:** `/ipns/base32(<HASH>)`
+
+Note: Base32 according to the [RFC4648](https://tools.ietf.org/html/rfc4648).
```
[the remark this answers is not in the document]
Comment (vasco-santos) [reply naming/README.md]: I will move this for now to the bottom of the document. We will see afterward if it makes sense to remove it!

**Comment by vasco-santos** [comment]
Thanks for the review @Stebalien 

I believe that I have taken all your feedback into account with the last commit. I changed the `Implementation Details` section to `Protocol` and modified the content accordingly.

File: naming/README.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+# Introduction
+
+Each time a file added to IPFS needs to be modified, its content addressable idendifier will be automatically modified. As a consequence, the address previously used for getting that file needs to be updated by who is using it. As this is not pratical, IPNS was conceptualized to solve the problem.
+
+IPNS is based on [SFS](http://en.wikipedia.org/wiki/Self-certifying_File_System). It consists of a PKI namespace, where a name is simply the hash of a public key. As a result, whoever controls the private key has full control over the name. Accordingly, records are signed by the private key and then distributed across the network (in IPFS, via the routing system). This is an egalitarian way to assign mutable names on the Internet at large, without any centralization whatsoever, or certificate authorities.
+
+# IPNS Record
+
+An IPNS record is a data structure composed by:
```
Comment (Stebalien) [inline naming/README.md]:
```suggestion
An IPNS record is a data structure containing the following fields:
```

("composed by" isn't really idiomatic)

File: naming/README.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  - [Introduction](#introduction)
+  - [IPNS Record](#ipns-record)
+  - [Protocol](#protocol)
+  - [Overview](#overview)
+  - [API Spec](#api-spec)
+  - [Integration with IPFS](#integration-with-ipfs)
+
+# Introduction
+
+Each time a file added to IPFS needs to be modified, its content addressable idendifier will be automatically modified. As a consequence, the address previously used for getting that file needs to be updated by who is using it. As this is not pratical, IPNS was conceptualized to solve the problem.
```
Comment (Stebalien) [inline naming/README.md]:
```suggestion
Each time a file is modified, its content address changes. As a consequence, the address previously used for getting that file needs to be updated by who is using it. As this is not pratical, IPNS was created to solve the problem.
```

Wording nits:

* IPFS doesn't react to need, unfortunately.
* IPNS is more than a concept (it exists).


"needs to be updated by who is using it" is still awkward but I'm not sure of a better wording off the top of my head.

File: naming/README.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+IPNS is based on [SFS](http://en.wikipedia.org/wiki/Self-certifying_File_System). It consists of a PKI namespace, where a name is simply the hash of a public key. As a result, whoever controls the private key has full control over the name. Accordingly, records are signed by the private key and then distributed across the network (in IPFS, via the routing system). This is an egalitarian way to assign mutable names on the Internet at large, without any centralization whatsoever, or certificate authorities.
+
+# IPNS Record
+
+An IPNS record is a data structure composed by:
+
+- 1. **Value** (bytes)
+  - It can be a path to another IPNS record, a `dnslink` path (`/ipns/example.com`) or an IPFS path (`/ipfs/Qm...`)
```
Comment (Stebalien) [inline naming/README.md]: We should make it clear that these are *just* examples. It can be *any* path.

File: naming/README.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+- 2. **Validity** (bytes)
+  - Expiration date of the record using [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) with nanoseconds precision.
+  - Note: Currently, the expiration date is the only available type of validity.
+- 3. **Validity Type** (uint64)
+  - Allows us to define if the record is valid under certain conditions
```
Comment (Stebalien) [inline naming/README.md]:
```suggestion
   - Allows us to define the conditions under which the record is valid.
```

File: naming/README.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  - Only supporting expiration date with `validityType = 0` for now
```
Comment (Stebalien) [inline naming/README.md]:
```suggestion
   - Only supports expiration date with `validityType = 0` for now.
```

**Comment by Stebalien** [comment]
LGTM. It looks ready to me but I'll let you merge it when you're ready.

**Comment by vasco-santos** [comment]
It is good to go! However, I do not have write access to merge this.

**Pull request merged**